### PR TITLE
Fix unchecked_bitwise_cast to "escape" its operand's ownership

### PR DESCRIPTION
**Pull request by atrick**
Fix unchecked_bitwise_cast to "escape" its operand's ownership

This makes it impossible for OSSA utilities to reason about the
lifetime of the value being cast. This fixes bugs in which SILGen
generates bitwise casts without protecting that lifetime.

Also fix LexicalDestroyHoisting to bail out when ownership escapes via a
`PointerEscape` operand.

Fixes rdar://100527903 (Overrelease of optional closure when using shorthand syntax)

The previous OSSA strategy was to gradually eliminate all "pointer
escape" SIL patterns. That would define away a class of bugs in the
OSSA utilities themselves. This was not friendly to developers working
on SILGen, and we still don't have verification that can catch these
bugs.

The new strategy is to make all OSSA utilities aware of pointer
escapes. This acknowledges reality that SIL is imperfect. Instead,
optimizer support for transforming SIL always needs to recognize
potentially dangerous patterns.

The downside of this new strategy is that it hides performance
problems because SIL optimizations can give up whenever they happen to
see a "pointer escape". We need to keep working on this problem
without being motivated by miscompiling code.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci benchmark

File: lib/SIL/IR/OperandOwnership.cpp [hunk 1]
```diff
@@ -238,9 +238,31 @@ OPERAND_OWNERSHIP(PointerEscape, ProjectExistentialBox)
 OPERAND_OWNERSHIP(PointerEscape, UncheckedOwnershipConversion)
 OPERAND_OWNERSHIP(PointerEscape, ConvertEscapeToNoEscape)
 
+// UncheckedBitwiseCast ownership behaves like RefToUnowned. It produces an
+// Unowned values from a non-trivial value, without consuming or borrowing the
```
Comment (nate-chandler) [inline lib/SIL/IR/OperandOwnership.cpp:242]: "an Unowned value"

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 2]
```diff
@@ -377,19 +377,26 @@ bool swift::findExtendedUsesOfSimpleBorrowedValue(
   return true;
 }
 
+// TODO: refactor this with SSAPrunedLiveness::computeLiveness.
 bool swift::findUsesOfSimpleValue(SILValue value,
                                   SmallVectorImpl<Operand *> *usePoints) {
   for (auto *use : value->getUses()) {
-    if (use->getOperandOwnership() == OperandOwnership::Borrow) {
+    switch (use->getOperandOwnership()) {
```
Comment (nate-chandler) [inline lib/SIL/Utils/OwnershipUtils.cpp:384]: Do we need the same change in LexicalDestroyFolding's findBorroweeUsage?

**Review by nate-chandler — APPROVED** [review]

File: lib/SIL/Utils/OwnershipUtils.cpp (hunk 2 above)
Comment (atrick) [reply lib/SIL/Utils/OwnershipUtils.cpp:384]: 👍 I'll need to figure out a test case for that one

**Comment by atrick** [comment]
@nate-chandler I'll merge this since CI is happy with it and the sooner it gets into builds the better. I'll immediately follow up with the destroy folding fix and comment grammar.

**Pull request merged**